### PR TITLE
Azure blob objectstore

### PR DESCRIPTION
**Pull request by alexlenail**
I've been speaking with @bgruening about this. For our purposes we need to use Azure to store our data, and this is a first step in that direction. This PR provides a new backend for the galaxy ObjectStore: Azure Blob Storage. The adapter was heavily modeled off of the S3 adapter in the same directory.

### CODE REVIEW

**Comment by alexlenail** [comment]
WARNING: At present, this PR is not ready to merge. Though it would do no immediate harm, I would like to test it further before merging.

**Comment by alexlenail** [comment]
@jxtx @nsoranzo I believe the checks began failing when I merged dev back into this branch. Travis says something about `flake8`. Can I assume this is not my problem?

**Comment by dannon** [comment]
@zfrenchee It does look like in the Travis output that the problematic file is `azureObjectstore.py` -- see the results here: https://travis-ci.org/galaxyproject/galaxy/jobs/145915553#L168

Let me know if you want help fixing this up, though, I'd be happy to give it a once-over and open a PR to your branch.

**Comment by alexlenail** [comment]
@dannon oh my gosh! I scrolled to the bottom much too fast. These seems like lint errors, let me look through them and see if I can fix them all in one pass. I'll save that review of yours for when I need it.

File: lib/galaxy/objectstore/azure_blob.py [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+import subprocess
+import threading
+import time
+
+from datetime import datetime
+
+from galaxy.exceptions import ObjectNotFound, ObjectInvalid
+from galaxy.util import umask_fix_perms, safe_relpath, directory_hash_id
+from galaxy.util.sleeper import Sleeper
+dfrom ..objectstore import ObjectStore, convert_bytes
```
Comment (nsoranzo) [inline lib/galaxy/objectstore/azure_blob.py]: Typo `dfrom`

Comment (alexlenail) [inline lib/galaxy/objectstore/azure_blob.py]: Sorry!

**Comment by alexlenail** [comment]
@dannon, Travis hasn't looked through the latest stuff here yet, but I think I'm ready for that first review. =)

**Comment by nsoranzo** [comment]
@zfrenchee https://travis-ci.org/galaxyproject/galaxy/jobs/146405694 , still failing unfortunately.

**Comment by alexlenail** [comment]
Hello @nsoranzo , @dannon :
Do I need to solve all the lint errors travis mentions? I believe the remaining lint errors are not issues. 

```
azure_blob.py:8:1: F401 'subprocess' imported but unused
azure_blob.py:23:5: F401 'azure.common.AzureConflictHttpError' imported but unused
azure_blob.py:23:5: F401 'azure.common.AzureMissingResourceHttpError' imported but unused
azure_blob.py:40:12: F821 undefined name 'BlobService'
azure_blob.py:40:31: E701 multiple statements on one line (colon)
azure_blob.py:60:5: E266 too many leading '#' for block comment
azure_blob.py:216:17: F841 local variable 'blob' is assigned to but never used
azure_blob.py:259:5: E266 too many leading '#' for block comment
azure_blob.py:487:5: E266 too many leading '#' for block comment
```

**Comment by dannon** [comment]
@zfrenchee At least the "BlobService" statement seems like it might be problematic.  Here's what I'd do:  https://github.com/dannon/galaxy/tree/minor_azure_tweaks

(I tried to open a PR on your fork, but for some reason I'm unable to)

(edit the second: github decided I was able to all of a sudden?   See https://github.com/zfrenchee/galaxy/pull/1)

**Comment by alexlenail** [comment]
@dannon @natefoo I believe you two are the experts on the ObjectStore. 

One thing you might notice about this PR is that all of the files which are stored as blobs end up being publicly accessible via a URL, which means "security by obscurity" is the only line of defense. You can imagine this wouldn't be tolerable for clinical data, for example. I am not entirely sure but I do believe ../s3.py objectStore plugin is the same way. I was wondering whether it would be possible to support data privacy with an objectstore plugin, which would be quite important for the group I belong to. I'm not sure exactly how to do this, and wanted to get this up and running for publicly available blobs first, but I think I'd like to be able to support that before merging this (though we can also do two merges). 

I'd love to get your thoughts and ideas about this.

**Comment by alexlenail** [comment]
Oh, and finally, I should mention that I still haven't figured out how to test this, mostly because I haven't quite figured out how configuration files work in galaxy. Specifically: I added an entry to `config/object_store_conf.xml.sample` : `<object_store type="Azure">` which has parameters like `<auth account_name="..." account_key="...." />`. How do those get filled in? Is it in galaxy.ini like [this page](https://wiki.galaxyproject.org/ObjectStore) seems to indicate (I think)? If so, why is none of the code on that page in galaxy.ini to start with?

(edit: jk I got it to work finally)

**Comment by dannon** [comment]
@zfrenchee I'll test this live on azure later today, and once I see how it works I'll try to answer the questions above.  Thanks for all the work so far!

(also, no need to keep merging galaxyproject/dev back into this -- it's fairly isolated and there won't be any problems merging)

**Comment by alexlenail** [comment]
@dannon I'm testing it a little bit now... and I might still not have the configuration piece down. When I try to upload, I get `AttributeError: 'NoneType' object has no attribute 'create'` where `create` is one of the methods on the `AzureObjectStore`. My `object_store_conf.xml` has `<object_store type="azure_blob" order="0">` but I have a feeling it's not actually building me an `AzureObjectStore` . What do you think?

**Comment by alexlenail** [comment]
How do you include packages in galaxy? Specifically, this PR requires that galaxy ship with

```
  from azure.storage import CloudStorageAccount
  from azure.storage.blob import BlockBlobService, PublicAccess
  from azure.common import AzureHttpError
```

**Comment by dannon** [comment]
@zfrenchee You'd want to add an entry in https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/dependencies/conditional-requirements.txt, and the configuration option to toggle it in https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/dependencies/__init__.py.

You'll also need to submit a pull request to galaxyproject/starforge for building the wheels for the azure package.

Once that's all in place, Galaxy will be able to grab the azure module automatically when necessary.

**Comment by alexlenail** [comment]
@dannon alright, I did the first part, you'll need to help me with starforge I suspect, because it seems a little over my head. Alternately, you can also just plug it in yourself. =)

**Comment by dannon** [comment]
@zfrenchee Created the starforge bit here:  https://github.com/galaxyproject/starforge/pull/102

**Comment by alexlenail** [comment]
@dannon 
How do I go about testing this now? I'm running into my own error: 

```
Exception: ObjectStore configured, but no azure.storage.blob dependency available.Please install and properly configure azure.storage.blob or modify Object Store configuration.
```

**Comment by alexlenail** [comment]
@dannon I've updated the wiki: https://wiki.galaxyproject.org/ObjectStore

**Comment by alexlenail** [comment]
@dannon still not sure how to test this, since I'm not sure how to get the dependencies fixed.

**Comment by dannon** [comment]
@zfrenchee I think you're probably running into the issue I've fixed here:  https://github.com/zfrenchee/galaxy/pull/2

**Comment by alexlenail** [comment]
Thanks for the PR -- that will make this more scalable in the longer term. I'm not sure that was the issue though =(.

**Comment by alexlenail** [comment]
Strangely enough, when I do `./run.sh` I see:

```
Requirement already satisfied (use --upgrade to upgrade): azure-storage in ./.venv/lib/python2.7/site-packages (from -r /dev/stdin (line 1))
Requirement already satisfied (use --upgrade to upgrade): azure-common in ./.venv/lib/python2.7/site-packages (from azure-storage->-r /dev/stdin (line 1))
```

So my guess is that there's something funky going on...

**Comment by alexlenail** [comment]
Aha -- I believe this may have something to do with the version: 

```
~/ » python
[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.72)] on darwin
>>> from azure.storage.blob import BlockBlobService
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name BlockBlobService
>>> from azure.storage.blob import BlobService
>>>
```

Azure storage recently moved from "BlobService" to "BlockBlobService". My PR assumes the newest azure sdk, but my bet is, like my computer, the sdk galaxy is trying to install is out of date.

**Comment by alexlenail** [comment]
yup that fixed it. Upgraded from version 0.20.0 to 0.32.0 (which is either the most recent or close to it)

**Comment by dannon** [comment]
@zfrenchee Great!  My guess is you had an older version installed (maybe via 'pip install azure' -- I think the version depended on there is 0.20.0) so Galaxy wasn't automatically installing the most up-to-date and compatible version in your .venv.

**Comment by alexlenail** [comment]
@dannon I have verified that this works! (and by that I mean that uploading data via the galaxy UI properly uploads data to MS azure). I think it's probably fair to remove the WIP label and discuss more rigorous testing strategies.

**Comment by bgruening** [comment]
@zfrenchee @dannon awesome work!
I removed the WIP label and added the review label.

**Comment by dannon** [comment]
@zfrenchee Don't worry about the new conflicts.  I'm going to test a few more things and will merge manually, resolving them.

**Comment by dannon** [comment]
I'm going to follow up with a separate PR to fix one bug I found, but this is nice and it's working fine for me.  Thanks @zfrenchee!

Merging locally to resolve conflicts.

**Comment by alexlenail** [comment]
@dannon if you aren't already exhausted by azure, I'm going to get started on another objectstore backend using another azure product they call "data lake". We at MIT will be using azure blobs in the short term, but may move to "data lake" in the medium term. Keep an eye out for that PR, and thanks for all the help so far =)

**Comment by natefoo** [comment]
> One thing you might notice about this PR is that all of the files which are stored as blobs end up being publicly accessible via a URL, which means "security by obscurity" is the only line of defense. You can imagine this wouldn't be tolerable for clinical data, for example. I am not entirely sure but I do believe ../s3.py objectStore plugin is the same way. I was wondering whether it would be possible to support data privacy with an objectstore plugin, which would be quite important for the group I belong to. I'm not sure exactly how to do this, and wanted to get this up and running for publicly available blobs first, but I think I'd like to be able to support that before merging this (though we can also do two merges).

@zfrenchee at least in the case of S3, aren't they private and only accessible because Galaxy has a secret key that allows access to the bucket?

**Comment by alexlenail** [comment]
Hi @natefoo 
I think at some point I figured out how to do this, and I agree with you about s3. I'll need to double check before we upload clinical data, but I think this Azure connector is now private by default.

**Pull request merged**